packages/vc-export/src/KiltAttestationProofV1.ts

```diff
@@ -0,0 +1,633 @@
+    throw new CredentialMalformedError('nonTransferable must be true')
+  // 5. check credentialStatus
+  if (credentialStatus.type !== KILT_REVOCATION_STATUS_V1_TYPE)
+    throw new CredentialMalformedError(
+      `credentialStatus must have type ${KILT_REVOCATION_STATUS_V1_TYPE}`
+    )
+  const { assetInstance, assetNamespace, assetReference } = Caip19.parse(
+    credentialStatus.id
+  )
+  const expectedAttestationId = credential.id.substring(
+    KILT_CREDENTIAL_IRI_PREFIX.length
+  )
+  if (
+    assetNamespace !== 'kilt' ||
+    assetReference !== 'attestation' ||
+    assetInstance !== expectedAttestationId
+  ) {
+    throw new Error(
+      `credentialStatus.id must end on 'kilt:attestation/${expectedAttestationId} in order to be verifiable with this proof`
+    )
+  }
+  // 6. json-ld expand credentialSubject
+  const expandedContents = jsonLdExpandCredentialSubject(credentialSubject)
+  // 7. Transform to normalized statments and hash
+  const { statements, digests } = normalizeClaims(expandedContents)
+  if (statements.length !== proof.salt.length)
+    throw new Error(
+      'Violated expectation: number of normalized statements === number of salts'
+    )
+  // 8-9. Re-compute commitments
+  const commitments = makeCommitments(
+    digests,
+    proof.salt.map((salt) => base58Decode(salt))
+  )
+  // 10. Assert commitments are in proof
+  commitments.forEach((recomputed, index) => {
+    if (!proof.commitments.includes(base58Encode(recomputed)))
+      throw new Error(
+        `No commitment for statement with digest ${u8aToHex(
+          digests[index]
+        )} and salt ${proof.salt[index]}`
+      )
+  })
+  // 11. Compute root hash
+  const rootHash = calculateRootHash(credential, proof)
+  // 12. Compare against credential id
+  if (credentialIdFromRootHash(rootHash) !== credential.id)
+    throw new Error('root hash not verifiable')
+
+  // 13. check that api is connected to the right network
+  const { api = ConfigService.get('api') } = opts
+  assertMatchingConnection(api, credential)
+  // 14. query info from chain
+  const {
+    cTypeId: onChainCType,
+    attester,
+    timestamp,
+    delegationId,
+  } = await verifyAttestedAt(rootHash, base58Decode(proof.block), { api })
+
+  if (attester !== issuer || onChainCType !== credential.credentialSchema.id) {
+    throw new Error(
+      `Credential not matching on-chain data: issuer "${attester}", CType: "${onChainCType}"`
+    )
+  }
+  // 16. Check timestamp
+  // 1-second precision should be enough
+  if (
+    Math.round(timestamp / 1000) !==
+    Math.round(new Date(credential.issuanceDate).getTime() / 1000)
+  )
+    throw new Error(
+      `block time ${new Date(
+        timestamp
+      ).toISOString()} does not match issuedAt (${credential.issuanceDate})`
+    )
+  // 17. + 18. validate federatedTrustModel items
+  await Promise.all(
+    credential.federatedTrustModel?.map(async (i) => {
+      if (i.type === KILT_ATTESTER_DELEGATION_V1_TYPE) {
```

Comment:
Perhaps this logic can be split into smaller chunks for easier readibility.

Reply:
Ah now I see what you mean! Alright, I think you'll like this: 1b0f5389163fe4beec78c7c40de06a7b72f764f1
(also realised that I wasn't properly checking the delegation node id in case there was no delegation-related item in the federatedTrust set)